src/variant_detection/snp_indel_detection.cpp

```diff
@@ -0,0 +1,94 @@
+#include <iostream>
+
+#include <seqan3/alphabet/cigar/cigar.hpp>
+#include <seqan3/core/debug_stream.hpp>
+#include <seqan3/io/sam_file/input.hpp>
+
+#include "iGenVar.hpp"
+#include "variant_detection/bam_functions.hpp"          // check the sam flags
+#include "variant_detection/snp_indel_detection.hpp"    // detect_snp_and_indel
+
+void detect_snp_and_indel(std::filesystem::path const & reads_filename)
+{
+    // Get the header information and set the necessary fields.
+    using sam_fields = seqan3::fields<seqan3::field::flag,       // 2: FLAG
+                                      seqan3::field::ref_id,     // 3: RNAME
+                                      seqan3::field::ref_offset, // 4: POS
+                                      seqan3::field::mapq,       // 5: MAPQ
+                                      seqan3::field::cigar>;     // 6: CIGAR
+
+    seqan3::sam_file_input reads_file{reads_filename, sam_fields{}};
+    auto const & header = reads_file.header();
+
+    // Check that the file is sorted.
+    if (header.sorting != "coordinate")
+        throw seqan3::format_error{"ERROR: Input file must be sorted by coordinate (e.g. samtools sort)"};
+
+    // Prepare one activity vector per reference sequence.
+    std::vector<std::vector<unsigned>> activity{};
+    activity.resize(header.ref_id_info.size()); // allocate the number of reference sequences
+    for (size_t idx = 0; idx < activity.size(); ++idx)
+        activity[idx].resize(std::get<0>(header.ref_id_info[idx])); // allocate the length of the reference sequence
+
+    for (auto && record : reads_file)
+    {
+        seqan3::sam_flag const flag = record.flag();                            // 2: FLAG
+        int32_t const ref_id        = record.reference_id().value_or(-1);       // 3: RNAME
+        int32_t ref_pos             = record.reference_position().value_or(-1); // 4: POS
+
+        // Skip reads with certain properties.
+        if (hasFlagUnmapped(flag) ||
+            hasFlagSecondary(flag) ||
+            hasFlagDuplicate(flag) ||
+            record.mapping_quality() < 20 ||
+            ref_id < 0 ||
+            ref_pos < 0)
+        {
+            continue;
+        }
+
+        // Track the current position within the read.
+        unsigned read_pos = 0;
+
+        // Step through the CIGAR string.
+        for (auto && [length, operation] : record.cigar_sequence())
+        {
+            // Skip long variants.
+            if (length >= 30) // TODO: use min_length parameter
```

Comment:
Since there are several of us working on this project, I have become accustomed to the following notation for TODOs:
```suggestion
            if (length >= 30) // TODO (joergi-w 30.09.2021): use min_length parameter
```
Then you know who to ask when you stumble across a TODO and also how old it is.